<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>SeqMiner Quick Guide</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}
</style>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: rgb(88, 72, 246)
   }

   pre .number {
     color: rgb(0, 0, 205);
   }

   pre .comment {
     color: rgb(76, 136, 107);
   }

   pre .keyword {
     color: rgb(0, 0, 255);
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: rgb(3, 106, 7);
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


</head>

<body>
<div id="toc">
<div id="toc_header">Table of Contents</div>
<ul>
<li>
<a href="#toc_0">SeqMiner Quick Guide</a>
<ul>
<li>
<a href="#toc_1">Extract gentoype matrix from VCF file</a>
</li>
<li>
<a href="#toc_2">Extract arbitrary fields from VCF file</a>
</li>
<li>
<a href="#toc_3">Applying rare-variant analysis</a>
</li>
<li>
<a href="#toc_4">Contact</a>
</li>
</ul>
</li>
</ul>
</div>


<h1 id="toc_0">SeqMiner Quick Guide</h1>

<p>SeqMiner(former name, vcf2geno) is a R package that helps you efficiently processing sequencing data in the <a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF format</a>.</p>

<p>SeqMiner requires VCF files indexed by Tabix. To use its gene based extraction, VCF need to be annotated by <a href="https://github.com/zhanxw/anno">SeqAnno</a>
We have provided an examplar VCF file in this package. You can use the following command to find its location on your computer:</p>

<pre><code class="r">library(vcf2geno)
fileName = system.file(&quot;vcf/all.anno.filtered.extract.vcf.gz&quot;, package = &quot;vcf2geno&quot;)
fileName
</code></pre>

<pre><code>## [1] &quot;/Library/Frameworks/R.framework/Versions/2.15/Resources/library/vcf2geno/vcf/all.anno.filtered.extract.vcf.gz&quot;
</code></pre>

<p>This manual demonstrates a quick guide on how to use SeqMiner.</p>

<h2 id="toc_1">Extract gentoype matrix from VCF file</h2>

<p>Extract genotype matrix from 1:196621007-196716634, and only focused on <em>Nonsynonymous</em> variants:</p>

<pre><code class="r">cfh &lt;- readVCFToMatrixByRange(fileName, &quot;1:196621007-196716634&quot;, &quot;Nonsynonymous&quot;)
</code></pre>

<pre><code>## Regcomp OK
</code></pre>

<pre><code class="r">cfh
</code></pre>

<pre><code>##             NA12286 NA12341 NA12342
## 1:196642233       1       2       0
## 1:196659237       0       2       1
</code></pre>

<p>In this example, we use the <em>fileName</em> for the VCF file path, <em>1:196621007-196716634</em> for the genomic range, and <em>Nonsynonymous</em> for the variants type.
Extraction results are stored in the variable <em>cfh</em>. Its row names are variant positions, while column names are sample IDs.</p>

<p>Similarly, we can extract genotype matrix per gene:</p>

<pre><code class="r">geneFile = system.file(&quot;vcf/refFlat_hg19.txt.gz&quot;, package = &quot;vcf2geno&quot;)
geneFile
</code></pre>

<pre><code>## [1] &quot;/Library/Frameworks/R.framework/Versions/2.15/Resources/library/vcf2geno/vcf/refFlat_hg19.txt.gz&quot;
</code></pre>

<pre><code class="r">cfh &lt;- readVCFToMatrixByGene(fileName, geneFile, &quot;CFH&quot;, &quot;Synonymous&quot;)
</code></pre>

<pre><code>## Regcomp OK
</code></pre>

<pre><code class="r">cfh
</code></pre>

<pre><code>##             NA12286 NA12341 NA12342
## 1:196654324       1       2       1
## 1:196682947       2       2       1
## 1:196654324       1       2       1
</code></pre>

<p>Note here we require a gene file in refFlat format. We have prepared several gene definition files in refFlat format using NCBI human build 37.
From this website, you can obtain these resources files.</p>

<p>The supported annotation variant type are: </p>

<pre><code># large strucutre variant
StructuralVariation

# indels
Stop_Gain
Stop_Loss
Start_Gain
Start_Loss
Frameshift        /* Indel length is not divisible by 3 */                                          
CodonGain         /* Insertion length is divisible by 3 */                                          
CodonLoss         /* Deletion length is divisible by 3 */                                           
CodonRegion       /* Just say the variant is in the Coding Region used in Structrual Varition*/
Insertion
Deletion

# single variants
Nonsynonymous
Synonymous
Essential_Splice_Site
Normal_Splice_Site
Utr5
Utr3
Exon
Intron
Upstream
Downstream
SNV           
Noncoding
Monomorphic
</code></pre>

<h2 id="toc_2">Extract arbitrary fields from VCF file</h2>

<p>For advanced data extraction from VCF file, there are two functions: <em>readVCFToListByRange</em> and <em>readVCFToListByGene</em>.
Their usage are similar to the previous two functions: <em>readVCFToMatrixByRange</em> and <em>readVCFToMatrixByGene</em> but the results are organized in list.</p>

<pre><code class="r">result &lt;- readVCFToListByGene(fileName, geneFile, &quot;CFH&quot;, &quot;Synonymous&quot;, c(&quot;CHROM&quot;, 
    &quot;ID&quot;), c(&quot;AC&quot;, &quot;AN&quot;), &quot;GT&quot;)
</code></pre>

<pre><code>## extractStringArray: [1] AC
## extractStringArray: [2] AN
## extractStringArray: [1] GT
## Regcomp OK
## Dump 9 elements:
## s[0] = &quot;0/1&quot;
## s[1] = &quot;1/1&quot;
## s[2] = &quot;0/1&quot;
## s[3] = &quot;1/1&quot;
## s[4] = &quot;1/1&quot;
## s[5] = &quot;0/1&quot;
## s[6] = &quot;0/1&quot;
## s[7] = &quot;1/1&quot;
## s[8] = &quot;0/1&quot;
</code></pre>

<pre><code class="r">result
</code></pre>

<pre><code>## $CHROM
## [1] &quot;1&quot; &quot;1&quot; &quot;1&quot;
## 
## $ID
## [1] &quot;.&quot; &quot;.&quot; &quot;.&quot;
## 
## $AC
## [1] &quot;4&quot; &quot;5&quot; &quot;4&quot;
## 
## $AN
## [1] &quot;6&quot; &quot;6&quot; &quot;6&quot;
## 
## $GT
##      [,1]  [,2]  [,3] 
## [1,] &quot;0/1&quot; &quot;1/1&quot; &quot;0/1&quot;
## [2,] &quot;1/1&quot; &quot;1/1&quot; &quot;1/1&quot;
## [3,] &quot;0/1&quot; &quot;0/1&quot; &quot;0/1&quot;
## 
## $sampleId
## [1] &quot;NA12286&quot; &quot;NA12341&quot; &quot;NA12342&quot;
</code></pre>

<p>There are 7 parameters in <em>readVCFToListByGene</em>: 
1. VCF file path
2. Gene definition file (refFlat foramt)
3. Character array of gene names (can be one gene or more than one gene names)
4. Annotation type
5. VCF field to extract (explain later)
6. VCF tags in the INFO field to extract (explain later)
7. VCF tags in the individual field to extract (explain later)</p>

<p>The last three parameter are closely related to VCF format.
In the VCF data lines, there are 8 fixed fields per record: CHROM, POS, ID, REF, ALT, QUAL, FILTER, INFO. These field, as well as the FORMAT field can be specifed in the 5th parameter. For example, specify c(&ldquo;CHROM&rdquo;, &ldquo;POS&rdquo;) will extract the genomic position.</p>

<p>In the VCF INFO field, there may be information including variant depth (DP), variant allele counts (AC), total allele counts (AN). To extract those information, use the 6th paramter. </p>

<p>In the VCF genotype fileds, which are from the 10th column to the last column per record, it contains the individual level information. Mostly commonly used information are genotype (GT), genotype depth (GD) and genotype quality (GQ). To extract these, specify c(&ldquo;GT&rdquo;, &ldquo;GD&rdquo;, &ldquo;GQ&rdquo;) in the 7th parameter.</p>

<p>The extraction results are stored in <em>result</em> in our example. All values are characters for simplicity. 
The extraction results is a list, and it contains CHROM, ID, AC, AN, GT and sampleId.
CHROM and ID are speicified in the 5th parameter; AC and AN are speicified by the 6th parameter; GT are specified in the last parameter.
We also list sample IDs in the sampleId part.</p>

<h2 id="toc_3">Applying rare-variant analysis</h2>

<p>SeqMiner can greatly reduce your effort to process huge sequencing files in VCF format. That indicates your valuable time can be spared on investigating causal variants.
Here we demonstrate how to apply <a href="http://www.hsph.harvard.edu/skat/">SKAT</a> method to study CFH gene. The genotype file we used is extracted from <a href="http://www.1000genomes.org/">1000 Genomes Project</a>. It contains 85 individuals in <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20111108_samples_pedigree/20111108_1000genomes_samples.xls">CEU population</a> and the variants are all located in the CFH gene region.  We will use simulated phenotype in this demonstration.</p>

<pre><code class="r"># Extract genotype matrix
vcf.1000g &lt;- system.file(&quot;vcf/1000g.phase1.20110521.CFH.anno.vcf.gz&quot;, package = &quot;vcf2geno&quot;)
geneFile = system.file(&quot;vcf/refFlat_hg19.txt.gz&quot;, package = &quot;vcf2geno&quot;)
geneName &lt;- &quot;CFH&quot;
cfh &lt;- readVCFToMatrixByGene(vcf.1000g, geneFile, geneName, &quot;Nonsynonymous&quot;)
</code></pre>

<pre><code>## Regcomp OK
</code></pre>

<pre><code class="r">dim(cfh)
</code></pre>

<pre><code>## [1] 47 85
</code></pre>

<pre><code class="r">
# Simulation phenotype under the null model
pheno &lt;- rnorm(ncol(cfh))

# Apply SKAT method
library(SKAT)
obj &lt;- SKAT_Null_Model(pheno ~ 1, out_type = &quot;C&quot;)
SKAT(t(cfh), obj)$p.value
</code></pre>

<pre><code>## Warning: 37 SNPs with either high missing rates or no-variation are
## excluded!
</code></pre>

<pre><code>## Warning: Genotypes of some variants are not the number of minor alleles!
</code></pre>

<pre><code>## [1] 0.2591
</code></pre>

<h2 id="toc_4">Contact</h2>

<p>SeqMiner is developed with joint effort from <a href="mailto:zhanxw@gmail.com">Xiaowei Zhan</a> and <a href="mailto:dajiang.liu@gmail.com">Dajiang Liu</a>.
We welcome your questions and feedbacks.</p>

</body>

</html>
